src/config.rs

```diff
@@ -2564,26 +2559,43 @@ impl TiKvConfig {
         self.resolved_ts.validate()?;
         self.resource_metering.validate()?;
 
-        let default_memory_usage_limit = Self::default_memory_usage_limit();
-        if self.memory_usage_limit.0 == 0 {
-            self.memory_usage_limit = default_memory_usage_limit;
-            return Ok(());
-        }
-        match self.memory_usage_limit.0.cmp(&default_memory_usage_limit.0) {
-            cmp::Ordering::Less => {
-                self.rocksdb
-                    .adjust_memory_usage_limit(self.memory_usage_limit);
-                self.raftdb
-                    .adjust_memory_usage_limit(self.memory_usage_limit);
-            }
-            cmp::Ordering::Greater => {
-                warn!(
-                    "memory_usage_limit {:?} is greater than total {:?}, fallback to total",
-                    self.memory_usage_limit, default_memory_usage_limit
-                );
-                self.memory_usage_limit = default_memory_usage_limit;
+        if self.memory_usage_limit.0.is_none() {
+            // Adjust `memory_usage_limit` if necessary.
+            if self.storage.block_cache.shared {
+                if let Some(cap) = self.storage.block_cache.capacity.0 {
+                    let limit = (cap.0 as f64 / BLOCK_CACHE_RATE * MEMORY_USAGE_LIMIT_RATE) as u64;
+                    self.memory_usage_limit.0 = Some(ReadableSize(limit));
+                } else {
+                    self.memory_usage_limit =
+                        OptionReadableSize(Some(Self::default_memory_usage_limit()));
+                }
+            } else {
+                let cap = self.rocksdb.defaultcf.block_cache_size.0
+                    + self.rocksdb.writecf.block_cache_size.0
+                    + self.rocksdb.lockcf.block_cache_size.0
+                    + self.raftdb.defaultcf.block_cache_size.0;
+                let limit = (cap as f64 / BLOCK_CACHE_RATE * MEMORY_USAGE_LIMIT_RATE) as u64;
+                self.memory_usage_limit.0 = Some(ReadableSize(limit));
             }
-            _ => {}
+        }
+
+        let mut limit = self.memory_usage_limit.0.unwrap();
+        let total = ReadableSize(SysQuota::memory_limit_in_bytes());
+        if limit.0 > total.0 {
+            warn!(
+                "memory_usage_limit:{:?} > total:{:?}, fallback to total",
+                limit, total,
+            );
+            self.memory_usage_limit.0 = Some(total);
```

Comment:
Is it proper to fall back to total?

Reply:
Do you mean we should fallback to `default`? I thinks some users maybe want a value between `default` and `total`. So here if it exceeds `total`, fallback to `total` is better than `default`.